### PR TITLE
Changes unarchive_dir and tests unarchive+prune algorithm

### LINKED ISSUE #2256: Avoid empty state while updating work directory in dynamic services
*Mark from control-core wrote*

You may recall our approach polls for a change in the file in input/input_1 every say .3 seconds. At startup, this file may not exist, and if there is a file exception, our code assumes a default value of zeros to start the simulation.  Later, our code assumes that *only* one of three things happens:  

1. the file is identical to the last time it was read (and so the polling continues)
1. the file's contents have changed (and so the polling stops) or
1. the file has zero bytes because the file has been reopened for write but has not yet been overwritten.   

This assumption is valid when we run the two nodes as processes on our local machine, but apparently not on OSPARC where it appears there is a brief moment when your system removes the old version of the file from input/input_1 before copying the new file from the other node.  This causes the default value of zeros to be inserted in the middle of a simulation, destroying the validity of the result.

### PR DESCRIPTION
**Pull request by pcrespov**
## What do these changes do?

- ``unarchive_dir`` returns set of paths extracted from archive 
- tests unarchive+prune algorithm will be used in the jupyter-commons library to fix #2256
   - ADDS ``PrunableFolder`` class to gather capture+prune operations

NOTE: These changes will be imported into jupyter-common next.

## Related issue/s

- FIXES: Avoid empty state while updating work directory in dynamic services #2256 
- Might FIX: Jupyter displays "Directory not found" menu permanently ITISFoundation/osparc-issues#412

## How to test

```
make devenv
source .venv/bin/activate
cd packages/service-library
make install-dev
make tests
```

## Checklist

- [ ] Use in jupyter-commons instead of copy&paste
- [ ]

### CODE REVIEW

File: packages/service-library/src/servicelib/archiving_utils.py [hunk 1]
```diff
@@ -88,7 +100,14 @@ async def unarchive_dir(archive_to_extract: Path, destination_folder: Path) -> N
...
             ]
 
-            await asyncio.gather(*tasks)
+            extracted_paths: List[Path] = await asyncio.gather(*tasks)
+
+            # NOTE: extracted_paths includes all tree leafs, which might include files and empty folders
+            return set(
+                p
+                for p in extracted_paths
+                if p.is_file() or (p.is_dir() and not any(p.glob("*")))
```
Comment (GitHK) [inline packages/service-library/src/servicelib/archiving_utils.py:109]: please rename p to something more readable at this point.

File: packages/service-library/tests/test_archiving_utils.py [hunk 2]
```diff
@@ -171,20 +160,47 @@ def _relative_path(input_path: Path) -> Path:
...
+        for f in src_dir.rglob("*")
+        if is_file_or_emptydir(f)
+    )
+    assert got_tails == expected_tails
 
 
 @pytest.mark.parametrize(
     "compress,store_relative_path",
-    [[True, True], [True, False], [False, True], [False, False]],
+    itertools.product([True, False], repeat=2),
```
Comment (GitHK) [inline packages/service-library/tests/test_archiving_utils.py:190]: 👍

File: packages/service-library/tests/test_archiving_utils.py [hunk 3]
```diff
@@ -171,20 +160,47 @@ def _relative_path(input_path: Path) -> Path:
...
+def assert_unarchived_paths(
+    unarchived_paths: Set[Path], src_dir: Path, dst_dir: Path, is_saved_as_relpath: bool
+):
+    is_file_or_emptydir = lambda p: p.is_file() or (p.is_dir() and not any(p.glob("*")))
+
+    # all unarchivedare under dst_dir
+    assert all(dst_dir in f.parents for f in unarchived_paths)
+
+    # can be also checked with strings
+    assert all(str(f).startswith(str(dst_dir)) for f in unarchived_paths)
```
Comment (GitHK) [inline packages/service-library/tests/test_archiving_utils.py:172]: maybe use path instead of f?

**Review by GitHK — APPROVED** [review]
Maybe consider some small renames

File: packages/service-library/src/servicelib/archiving_utils.py (hunk 1 above)
Comment (pcrespov) [reply packages/service-library/src/servicelib/archiving_utils.py:109]: I find this very readable :-)   if "is a file or a directory with nothing" maps nicely, oder? :-)

File: packages/service-library/tests/test_archiving_utils.py (hunk 3 above)
Comment (pcrespov) [reply packages/service-library/tests/test_archiving_utils.py:172]: line 172 is checking exactly the same as line 169 but "the old way" . In fact line 169 is more reliable ... I was just comparing

**Review by sanderegg — APPROVED** [review]
pair reviewed

**Pull request merged**